### ISSUE TITLE
SVD intermittently reports incorrect tensor size

### ISSUE THREAD

**Issue by rustrust** [LaurentMazare/tch-rs#393]
I need to check whether a matrix I am working with is singular, so I am placing a call to linalg_matrix_rank. Everything is fine when the matrix is full rank, but tch-rs panics if the matrix is at all singular.

Since the purpose of this function is to report on the rank of the matrix, which is not necessarily full rank, some other behavior is preferred. Most preferably, this function should never panic or error if the matrix is singular, because its purpose is to report on the numeric rank of the matrix. 

```
let rank = my_tensor.reshape(&[1, 1, 1024, 1024]).linalg_matrix_rank(None, false);
println!("my_tensor linalg_matrix_rank: {:?}", &rank);

let rank2 = my_tensor.reshape(&[1, 1, 1024, 1024]).matrix_rank(false);
println!("my_tensor matrix_rank: {:?}", &rank2);
```

The result of running this looks like:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Torch("svd_cuda: For batch 0: U(1025,1025) is zero, singular U.\nException raised from batchCheckErrors at /pytorch/aten/src/ATen/native/LinearAlgebraUtils.h:251
```

**Comment by LaurentMazare** [comment]
Looks like this error is raised from [here](https://github.com/pytorch/pytorch/blob/1ce3281a6d266d627f75a283dac5ce10c9f4a27b/aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu#L508). It's a bit hard to understand why this is checked but note that tch-rs is just a thin wrapper around the C++ api of PyTorch here so the exception raised on the C++ side is converted to a Rust `Result`.
If you don't want this call to panic, you can use the *fallible* versions of the functions, e.g. `f_matrix_rank` and `f_linalg_matrix_rank` and check whether the output is an error or not.
If you think that these functions should never fail, I would suggest reporting the issue upstream to the PyTorch repo, the best for that would be to first replicate the issue in C++ so that it's easier for PyTorch devs to have a look.

**Comment by rustrust** [comment]
f_linalg_matrix_rank does Err on this, which is nice. The error message, however, is a page of text and doesn't include the matrix rank. I will file an issue with pytorch and reference this from there.

The error message begins (only pasting a snippet here):
```
Err(Torch("svd_cuda: For batch 0: U(1025,1025) is zero, singular U.\nException raised from batchCheckErrors at /pytorch/aten/src/ATen/native/LinearAlgebraUtils.h:251 (most recent call first):\nframe #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x69 (0x7f6259f4f1d9 in ...
```

**Comment by rustrust** [comment]
I am encountering this same problem when I am calculating only SVD via tch::Tensor::svd using:
```
 let (u, s, v) = my_tensor.reshape(&[1, 1, 1024, 1024]).svd(true, true);
```

The PyTorch folks pointed out something I didn't notice, which is that the SVD is saying that U is (1025,1025), but you can see that I am reshaping the matrix to be of dimension 1024 by 1024. 

I am encountering this error when I add a value I am calculating to the loss function of the convolutional neural network mnist example.

I have added no lines of code which need to be marked unsafe and I am not modifying the dimensions of any tensors. I am not spinning up any extra threads.

I am carefully printing out the size of the tensor on which I am calculating the SVD, immediately before the tch::Tensor::svd call and it is always 1024 by 1024 (and you can see that the reshape enforces this), including during the iteration immediately preceding this error. I also calculate the size of u after calculating the svd (on the iterations where svd does not fail), and it is always 1024x1024.

my_tensor points at the weights of  one of the neural network layers.

**Comment by LaurentMazare** [comment]
The best would probably be to make a repro in C++ so that PyTorch devs could comment (and ideally the repro should be as short as possible). It's very likely that what you do on the rust side could be done with the C++ api as these are basically the same.

If it's hard to get a C++ repro, having an as short as possible Rust repro could be interesting too, and maybe someone else can look at converting it to C++.

**Comment by rustrust** [comment]
Start with the mnist/mnist_conv.rs file, copy it into a new example which we can refer to as mnist_conv_svdbug.rs

You may need a little bit of patience. If the bug does not appear within a couple of epochs, I abort the run with ctrl+C, rereun.

Make the following few modifications. Overall we are convolving fc1 with a small template, reducing that to a scalar, and then including that scalar in the loss function.

This goes above the epoch for loop, just after the definition of net:
```
let my_template = Tensor::of_slice(&[1.0_f32, -1.0_f32, 1.0_f32, -1.0_f32, 1.0_f32, -1.0_f32, 1.0_f32, -1.0_f32, 1.0_f32])
        .to_device(Device::Cuda(0))
        .reshape(&[1, 1, 3, 3]);

let mut my_template_fc1 = net.fc1.ws.reshape(&[1, 1, 1024, 1024]);
```

Then, within innermost for loop over (bimages, blabels), add this code. Here we calculate the convolution of the template with the fc1 layer of the convnet and then calculate its magnitude. (In the next code block we will include the magnitude with the loss function.)

```
            my_template_fc1 = net.fc1.ws
                  .reshape(&[1, 1, 1024, 1024])
                  .conv2d(&my_template, None::<Tensor>,
                                        &[1], // stride
                                        &[0], // padding
                                        &[1], // dilation
                                        1     // groups
                                        );

//            let my_template_sum_value = -my_template_fc1.abs().sum(Kind::Float); // calculating my_template_sum_value this way does not appear to surface the bug

            let my_template_sum_value = -(my_template_fc1.multiply(&my_template_fc1) + my_template_fc1.multiply(&my_template_fc1)).sqrt().sum(Kind::Float); // calculating my_template_sum_value like this results in SVD claiming U is 1025x1025 within roughly a few dozen to a couple hundred iterations

                match net.fc1.ws.reshape(&[1, 1, 1024, 1024]).f_svd(true, true)
            {
                Ok(a) => { println!("fc1 svd ok"); },
                Err(e) => {
                   println!("################ fc1 svd fail #######################################");
                    println!("{:?}", e);
                    panic!("we are done here");
                }

```

The loss function should have a multiple of this scalar added, like so:

```
            let loss = net
                .forward_t(&bimages, true)
                .cross_entropy_for_logits(&blabels) + 0.5*my_template_sum_value;
```

The value multiplying the scalar (in this case, 0.5) appears to influence how regularly the U1025 svd bug appears.

I compile like so:
```
RUST_BACKTRACE=1 cargo run --release --example mnist conv_svdbug
```

```
rustc --version
rustc 1.54.0 (a178d0322 2021-07-26)

nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2021 NVIDIA Corporation
Built on Wed_Jun__2_19:15:15_PDT_2021
Cuda compilation tools, release 11.4, V11.4.48
Build cuda_11.4.r11.4/compiler.30033411_0

libtorch 1.9.0+cu111

```

**Comment by LaurentMazare** [comment]
Closing this as it has been a while and it seems that the underlying pytorch issue has been closed [pytorch#62166](https://github.com/pytorch/pytorch/issues/62166). Feel free to re-open if it's still a problem.
